### PR TITLE
feat(context-agent): tool status callbacks and process support

### PR DESCRIPTION
**Pull request by abeatrix**
CLOSE https://linear.app/sourcegraph/issue/CODY-4280 & https://linear.app/sourcegraph/issue/CODY-4282

Screenshot 2024-12-27 at 12 27 50 AM

This change introduces the following improvements to the Cody tool system:

1. Added support for displaying a loading indicator when the context is being loaded as shown in [figma](https://www.figma.com/design/f078wFMKsIOaEwj7Iwj5xy/Unified-Cody?node-id=7560-18281&t=Un5MlNVhg9RBCImr-0)
2. Updated the step message item to display the step ID and content in a more compact and readable format.
3. Added support for displaying an error icon when a step fails.
4. Adjusted the styling and layout of the step message items to improve the overall visual appearance.
5. Added a 30-second timeout for tool execution to prevent indefinite blocking.
6. Implemented a `Promise.race` between the tool execution and the timeout promise to ensure timely completion.
7. Updated the `DeepCodyAgent` to handle tool execution errors and return an empty result set instead of propagating the error.

## Test plan

See demo:

https://github.com/user-attachments/assets/57ac1f7d-a596-4671-bfa1-3ffd1611e327

Update storybook:

image

### CODE REVIEW

File: vscode/src/chat/agentic/CodyChatAgent.ts [hunk 1]
```diff
@@ -5,24 +5,30 @@ import {
     type Message,
     type PromptMixin,
     type PromptString,
+    errorToChatError,
     newPromptMixin,
 } from '@sourcegraph/cody-shared'
+import type { ChatMessageStep } from '@sourcegraph/cody-shared/src/chat/transcript/messages'
```
Comment (valerybugakov) [inline vscode/src/chat/agentic/CodyChatAgent.ts:11]: Let's add it to the root export folder of the `@sourcegraph/cody-shared` package and import it together with other items from it.

File: lib/shared/src/chat/transcript/messages.ts [hunk 2]
```diff
@@ -39,6 +39,15 @@ export interface ChatMessage extends Message {
...
     manuallySelectedIntent?: 'search' | 'chat' | 'edit' | 'insert' | undefined | null
     search?: ChatMessageSearch | undefined | null
+    steps?: ChatMessageStep[] | undefined | null
+}
+
+export interface ChatMessageStep {
+    id: string
+    content: string
+    status: 'pending' | 'success' | 'error'
+    step?: number
```
Comment (valerybugakov) [inline lib/shared/src/chat/transcript/messages.ts:49]: Do we use this field anywhere? What's its purpose?

File: vscode/src/chat/agentic/CodyChatAgent.ts [hunk 3]
```diff
@@ -80,4 +86,59 @@ export abstract class CodyChatAgent {
...
+
+    public setStatusCallback(postMessage: (model: string) => void): void {
+        this.postMessageCallback = postMessage
+
+        const model = this.chatBuilder.selectedModel ?? ''
+        let steps = this.chatBuilder.getLastMessageSteps() ?? []
+
+        const createStep = (
+            content: string,
+            id: string,
```
Comment (valerybugakov) [inline vscode/src/chat/agentic/CodyChatAgent.ts:98]: Would `toolName` be more accurate here?

File: vscode/src/chat/agentic/CodyChatAgent.ts [hunk 4]
```diff
@@ -80,4 +86,59 @@ export abstract class CodyChatAgent {
...
+            stepNum: number,
+            status: ChatMessageStep['status']
+        ) => ({
```
Comment (valerybugakov) [inline vscode/src/chat/agentic/CodyChatAgent.ts:101]:
```suggestion
        ): ChatMessageStep => ({
```

File: vscode/src/chat/agentic/CodyChatAgent.ts [hunk 5]
```diff
@@ -80,4 +86,59 @@ export abstract class CodyChatAgent {
...
+            content,
+            id,
+            step: stepNum,
+            status,
+        })
+
+        this.statusCallback = {
+            onToolsStart: () => {
+                // Initialize steps array with an empty pending step
+                steps = [createStep('', '', 0, 'pending')]
```
Comment (valerybugakov) [inline vscode/src/chat/agentic/CodyChatAgent.ts:111]: Argument list consisting of `'', '', 0,` can be tricky to read. Can we do anything to make it easier for readers?

File: vscode/src/chat/agentic/CodyChatAgent.ts [hunk 6]
```diff
@@ -32,7 +38,7 @@ export abstract class CodyChatAgent {
         for (const [tag, tool] of this.toolHandlers) {
             this.multiplexer.sub(tag, {
                 onResponse: async (content: string) => tool.stream(content),
-                onTurnComplete: async () => tool.stream(''),
+                onTurnComplete: async () => {},
```
Comment (valerybugakov) [inline vscode/src/chat/agentic/CodyChatAgent.ts:41]: Curious, why is this change needed?

File: vscode/src/chat/agentic/CodyTool.ts [hunk 7]
```diff
@@ -86,7 +90,33 @@ export abstract class CodyTool {
...
+                    setTimeout(() => {
+                        reject(
+                            new Error(
+                                `${this.config.title} execution timed out after ${CodyTool.EXECUTION_TIMEOUT_MS}ms`
+                            )
+                        )
+                    }, CodyTool.EXECUTION_TIMEOUT_MS)
+                })
+                // Race between execution and timeout
+                const results = await Promise.race([this.execute(span, queries), timeoutPromise])
```
Comment (valerybugakov) [inline vscode/src/chat/agentic/CodyTool.ts:110]: Nice one!

File: vscode/src/chat/agentic/DeepCody.ts [hunk 8]
```diff
@@ -131,8 +135,19 @@ export class DeepCodyAgent extends CodyChatAgent {
...
-                Array.from(this.toolHandlers.values()).map(tool => tool.execute(span))
+                Array.from(this.toolHandlers.entries()).map(async ([name, tool]) => {
+                    // Check abort signal before each tool run
+                    if (chatAbortSignal.aborted) {
+                        return []
+                    }
+                    try {
+                        return await tool.run(span, this.statusCallback)
+                    } catch (error) {
+                        return []
```
Comment (valerybugakov) [inline vscode/src/chat/agentic/DeepCody.ts:146]: The timeout error would be thrown here from the newly added `Promise.race` logic. Do we want to track timeouts and add some analytics logging to understand how often it happens?

**Review by valerybugakov — COMMENTED** [review]
Went through the code and left some comments inline. Will test the UI changes locally a bit later today!

**Review by valerybugakov — COMMENTED** [review]
The updated UI looks really nice! The contrast of checkmarks can be improved for the light theme:

Screenshot 2024-12-27 at 20 56 58

**Review by valerybugakov — APPROVED** [review]
Approving to unblock ✅

File: vscode/src/chat/agentic/CodyChatAgent.ts (hunk 6 above)
Comment (abeatrix) [reply vscode/src/chat/agentic/CodyChatAgent.ts:41]: because the previous code wasn't really doing anything since it just send an empty string 😅

File: lib/shared/src/chat/transcript/messages.ts (hunk 2 above)
Comment (abeatrix) [reply lib/shared/src/chat/transcript/messages.ts:49]: ahh i should add some docstrings for each field. Thanks!
